back/src/oauth2.ts

```diff
@@ -0,0 +1,143 @@
+    const grant = await prisma.createGrant({
+      user: { connect: { id: user.id } },
+      code: getUid(16),
+      application: { connect: { id: client.id } },
+      expires: 10 * 60, // recommended by https://tools.ietf.org/html/rfc6749#section-4.1.2
+      redirectUri
+    });
+    done(null, grant.code);
+  })
+);
+
+export const tokenErrorMessages = {
+  invalid_code: "Invalid authorization code",
+  invalid_redirect_uri: "Invalid redirect uri",
+  invalid_client_id: "Invalid client id",
+  grant_expired: "Grant has expired"
+};
+
+// Exchange authorization codes for access tokens. The callback accepts the
+// `client`, which is exchanging `code` and any `redirectUri` from the
+// authorization request for verification. If these values are validated, the
+// application issues an access token on behalf of the user who authorized the
+// code. The issued access token response can include a refresh token and
+// custom parameters by adding these to the `done()` call
+oauth2server.exchange(
+  oauth2orize.exchange.code(async (client, code, redirectUri, done) => {
+    const fragment = `fragment GrantWithUserAndApplication on Grant {
+      createdAt
+      code
+      expires
+      redirectUri
+      user { id, name, email }
+      application { id }
+    }`;
+
+    type GrantWithUserAndApplication = Grant & { user: User } & {
+      application: Application;
+    };
+
+    const grant = await prisma
+      .grant({ code })
+      .$fragment<GrantWithUserAndApplication>(fragment);
+    if (!grant) {
+      const err = new TokenError(
+        tokenErrorMessages.invalid_code,
+        "invalid_grant"
+      );
+      return done(err);
+    }
+    if (redirectUri != grant.redirectUri) {
+      const err = new TokenError(
+        tokenErrorMessages.invalid_redirect_uri,
+        "invalid_grant"
+      );
+      return done(err);
+    }
+    if (client.id != grant.application.id) {
+      const err = new TokenError(
+        tokenErrorMessages.invalid_client_id,
+        "invalid_grant"
+      );
+      return done(err);
+    }
+    if (isExpired(grant)) {
+      const err = new TokenError(
+        tokenErrorMessages.grant_expired,
+        "invalid_grant"
+      );
+      return done(err);
+    }
+
+    const token = getUid(40);
+
+    const accessToken = await prisma.createAccessToken({
+      user: {
+        connect: { id: grant.user.id }
+      },
+      application: {
+        connect: { id: grant.application.id }
+      },
+      token
+    });
+
+    // Add custom params
+    let params = { user: { name: grant.user.name, email: grant.user.email } };
+
+    return done(null, accessToken.token, null, params);
+  })
+);
+
+/**
+ * Check whether or not a grant has expired
+ * @param grant
+ */
+export function isExpired(grant: Grant): boolean {
+  // use new Date(Date.now()) instead of new Date() in order to mock function in test
+  const now = new Date(Date.now());
```

Comment:
Et on peut pas se passer du `new Date()` autour du coup ? Et ainsi éviter le `getTime()` dessus après ?
Juste pour éviter le `new Date(Date.now());` qui m'a surpris à la lecture (je n'ai vu le comment qu'après)
```suggestion
  const now = Date.now();
  const elasped = (now - createdAt.getTime()) / 1000;
```

Reply:
ok c'est corrigé